### DISCUSSION TITLE
How to test browser code under Deno

### DISCUSSION THREAD

**Discussion by David-Else** [denoland/deno#11473] [Q&A]
`deno test --config ./tsconfig.json --watch --allow-all --no-check test/ ` (or without `--no-check`) returns:

```
error: Uncaught ReferenceError: document is not defined
const elem = document.getElementById("test") as HTMLElement;
             ^
```
`tsconfig.json`
```
{
  "compilerOptions": {
    "target": "esnext",
    "lib": ["dom", "dom.iterable", "dom.asynciterable", "deno.ns"]
  }
}
```

`main.ts`
```ts
const elem = document.getElementById("test") as HTMLElement;
```
The LSP picks up my `tsconfig.json` fine and accepts `document` when use `config = "./tsconfig.json"`, and `test` is reading the file as it says:
```
Unsupported compiler options in "/home/david/Documents/apps-sites/typeScript/dependency-heaven-
typescript/tsconfig.json".
  The following options were ignored:
    target
```

I have a test repo here to show the problem: https://github.com/David-Else/dependency-heaven-typescript

**Comment by nayeemrmn** [comment]
That's a runtime error, which is why it's happening even with `--no-check`.

**Comment by David-Else** [comment]
@nayeemrmn It happens without `--no-check`:
```
deno test --config ./tsconfig.json --watch --allow-all test/
Check file:///home/david/Documents/apps-sites/typeScript/dependency-heaven-typescript/test/mo
d_test.ts
Unsupported compiler options in "/home/david/Documents/apps-sites/typeScript/dependency-heave
n-typescript/tsconfig.json".
  The following options were ignored:
    target
Check file:///home/david/Documents/apps-sites/typeScript/dependency-heaven-typescript/test/mo
d_test.ts
Unsupported compiler options in "/home/david/Documents/apps-sites/typeScript/dependency-heave
n-typescript/tsconfig.json".
  The following options were ignored:
    target

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (1497ms)

error: Uncaught ReferenceError: document is not defined
const elem = document.getElementById("test") as HTMLElement;
             ^
```

**Comment by nayeemrmn** [comment]
Yes, because it's a runtime error. There is no `document` global in Deno. The `"dom"` lib to bypass the type-checking error isn't going to change that. What's your expected behaviour?

**Comment by David-Else** [comment]
Sorry, I am struggling a bit. I would like to test my browser code using Deno, is there any information available on how this is done?

I have found `import jsdom from "https://cdn.skypack.dev/jsdom";`, and https://github.com/b-fuze/deno-dom ... is there any guide I could follow or best practice? Not sure where to start. Thanks!

**Comment by caspervonb** [comment]
> Sorry, I am struggling a bit. I would like to test my browser code using Deno, is there any information available on how this is done?

Tangental, but [I'm releasing a complementary toolchain _soon_](https://twitter.com/caspervonb/status/1415690606983090184) which shims the Deno test harness and runs the tests in browsers.

**Comment by krutoo** [comment]
Is there some news for Deno 2.0? I want to use happy-dom with deno
